# 6.1 Sourcing Open Data

### This script includes the following points:

### 1. Data Information and Basic Statistics
### 2. Missing Values
### 3. Data Types
### 4. Duplicates

In [1]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Import Data
path = r"D:\Users\Marlo\OneDrive\OneDrive - South Washington County Schools\Home\job support\Coursework\Achievement 6\Disability Statistics"

In [3]:
prev = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Disability Prevalence.csv'))

In [4]:
employ = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Employment & Poverty.csv'))

In [5]:
earn = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Earnings.csv'))

## 1. Data Information and Basic Statistics

In [6]:
#Info on disability prevalence data set
prev.head()

state  total_pop disability_category age_group race gender  \
0    US  326942778                 All       All  All    All   
1    AL    4956828                 All       All  All    All   
2    AK     703612                 All       All  All    All   
3    AZ    7175530                 All       All  All    All   
4    AR    2975548                 All       All  All    All   

   disability_count  percent_of_total  percent_of_disabilities  
0        42601999.0              13.0                      NaN  
1          805849.0              16.3                      NaN  
2           99194.0              14.1                      NaN  
3          970430.0              13.5                      NaN  
4          517723.0              17.4                      NaN

In [7]:
prev.shape

(4030, 9)

In [8]:
prev.describe()

total_pop  disability_count  percent_of_total  \
count  4.030000e+03      4.014000e+03       4014.000000   
mean   2.320673e+06      1.324406e+05          5.534554   
std    6.341655e+06      7.120808e+05          5.250360   
min    4.188000e+03      3.200000e+01          0.100000   
25%    2.818380e+05      8.133250e+03          2.300000   
50%    8.670870e+05      3.831850e+04          4.200000   
75%    2.820287e+06      1.334220e+05          6.500000   
max    3.269428e+08      4.260200e+07         41.300000   

       percent_of_disabilities  
count              3401.000000  
mean                 31.391355  
std                  15.614563  
min                   0.000000  
25%                  18.600000  
50%                  30.000000  
75%                  40.500000  
max                 100.000000

In [9]:
#Info on employment & poverty data set
employ.head()

state  total_pop disability_category race gender  employed_count  \
0    AL     427044                 All  All    All        138077.0   
1    AK      57446                 All  All    All         26466.0   
2    AZ     462266                 All  All    All        203672.0   
3    AR     270410                 All  All    All         95322.0   
4    CA    2094680                 All  All    All        825876.0   

   disability_employed  no_disability_employed  employment_gap  poverty_count  \
0                 32.3                    74.6            42.3       117608.0   
1                 46.1                    74.9            28.8        13785.0   
2                 44.1                    76.0            31.9       101643.0   
3                 35.3                    76.8            41.5        75911.0   
4                 39.4                    73.5            34.1       470724.0   

   disability_poverty  no_disability_poverty  poverty_gap  
0                27.5                   14.4         13.1  
1                24.0                    9.1         14.9  
2                22.0                   12.1          9.9  
3                28.1                   14.4         13.7  
4                22.5                   11.3         11.2

In [10]:
employ.shape

(714, 13)

In [11]:
employ.describe()

total_pop  employed_count  disability_employed  \
count  7.140000e+02      710.000000           709.000000   
mean   1.427106e+05    53556.743662            39.425106   
std    2.058733e+05    81685.193259            13.563442   
min    9.400000e+01       26.000000             7.600000   
25%    2.258775e+04     8520.750000            30.600000   
50%    7.033400e+04    24821.500000            40.100000   
75%    1.823118e+05    66670.750000            47.900000   
max    2.094680e+06   832902.000000            95.000000   

       no_disability_employed  employment_gap  poverty_count  \
count              408.000000      403.000000     710.000000   
mean                76.457108       33.146650   57619.433803   
std                  4.894689        9.706922   74003.606530   
min                 54.900000      -18.000000      13.000000   
25%                 73.600000       29.250000    9717.000000   
50%                 76.400000       34.500000   31155.500000   
75%                 79.700000       38.800000   78076.000000   
max                 90.900000       70.100000  470724.000000   

       disability_poverty  no_disability_poverty  poverty_gap  
count          710.000000             408.000000   404.000000  
mean            23.171127              14.866667     9.026485  
std              7.454352               5.935382     7.468924  
min              2.000000               4.800000   -25.600000  
25%             18.825000              10.800000     6.100000  
50%             22.600000              13.300000     8.600000  
75%             26.800000              17.400000    11.500000  
max            100.000000              47.500000    85.100000

In [12]:
#Info on earnings data set
earn.head()

state  total_earnings  no_disability_earnings  disability_earnings  \
0    AL         44212.0                 44278.0              40266.0   
1    AK         58319.0                 59291.0              53857.0   
2    AZ         51055.0                 51404.0              43242.0   
3    AR         41172.0                 41192.0              33688.0   
4    CA         60861.0                 60970.0              51427.0   

   earnings_gap  SSDI_allowance_rate  min_wage_rate  Unnamed: 7  Unnamed: 8  
0        4012.0                 35.3           7.25         NaN         NaN  
1        5434.0                 57.7          10.34         NaN         NaN  
2        8162.0                 35.7          12.80         NaN         NaN  
3        7504.0                 38.7          11.00         NaN         NaN  
4        9543.0                 36.0          14.00         NaN         NaN

In [13]:
earn.shape

(59, 9)

There are only 51 rows and 7 columns in this data set. Will need to fix later on.

In [14]:
earn.describe()

total_earnings  no_disability_earnings  disability_earnings  \
count       51.000000               51.000000            51.000000   
mean     53042.058824            53520.196078         44875.372549   
std       8836.202543             8906.085838          7136.892808   
min      40890.000000            40912.000000         33307.000000   
25%      46332.500000            46765.500000         40713.500000   
50%      51400.000000            51413.000000         43228.000000   
75%      57069.500000            58445.500000         49509.000000   
max      91860.000000            92266.000000         71474.000000   

       earnings_gap  SSDI_allowance_rate  min_wage_rate  Unnamed: 7  \
count     51.000000            51.000000      51.000000         0.0   
mean    8644.823529            41.200000       9.987059         NaN   
std     4010.603584             5.532956       2.637969         NaN   
min    -2639.000000            32.500000       7.250000         NaN   
25%     5599.000000            36.950000       7.250000         NaN   
50%     8837.000000            41.100000       9.950000         NaN   
75%    10294.000000            44.550000      12.375000         NaN   
max    20792.000000            57.700000      16.100000         NaN   

       Unnamed: 8  
count         0.0  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

## 2. Missing Values

In [15]:
#Missing Values for disability prevalence data set
prev.isnull().sum()

state                        0
total_pop                    0
disability_category          0
age_group                    0
race                         0
gender                       0
disability_count            16
percent_of_total            16
percent_of_disabilities    629
dtype: int64

- 16 missing values in the columns disability_count, percent_of_total, and percent_of_disabilities are due to no information being available and will remain blank for that reason.
- 607 missing values in the column percent_of_disabilities is due to these rows being information about all disabilities so it cannot be compared to all disabilities. They will stay blank as a result.

In [16]:
#Missing Values for employment & poverty data set
employ.isnull().sum()

state                       0
total_pop                   0
disability_category         0
race                        0
gender                      0
employed_count              4
disability_employed         5
no_disability_employed    306
employment_gap            311
poverty_count               4
disability_poverty          4
no_disability_poverty     306
poverty_gap               310
dtype: int64

- There are 4-5 missing values in the columns employed_count, disability_employed, employment_gap, poverty_count, disability_poverty, and poverty_gap due to no information being available and will remain blank for that reason.
- 306 missing values in the columns no_disability_employed, employment_gap, no_disability_poverty, and poverty_gap is dues to these rows being information about one specific disability category so cannot be compared to all people without disabilities. They will stay blank as a result.

In [17]:
#Missing Values for earnings data set
earn.isnull().sum()

state                      8
total_earnings             8
no_disability_earnings     8
disability_earnings        8
earnings_gap               8
SSDI_allowance_rate        8
min_wage_rate              8
Unnamed: 7                59
Unnamed: 8                59
dtype: int64

2 empty columns and 8 empty rows were added to this data set. They need to be deleted.

In [18]:
#Delete 2 extra columns and 8 extra rows in earnings data set.

In [19]:
earn = earn.drop('Unnamed: 7', axis = 'columns')

In [20]:
earn = earn.drop('Unnamed: 8', axis = 'columns')

In [21]:
earn.shape

(59, 7)

In [22]:
earn3 = earn.drop(index=earn.index[-8:])

In [23]:
earn3.shape

(51, 7)

In [24]:
earn = earn3

In [25]:
earn.shape

(51, 7)

In [26]:
#Missing Values for earnings data set
earn.isnull().sum()

state                     0
total_earnings            0
no_disability_earnings    0
disability_earnings       0
earnings_gap              0
SSDI_allowance_rate       0
min_wage_rate             0
dtype: int64

There are no missing values in the earnings data set.

## 3. Data Types

In [27]:
#Check Data Types in disability prevalence dataset
prev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4030 entries, 0 to 4029
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   state                    4030 non-null   object 
 1   total_pop                4030 non-null   int64  
 2   disability_category      4030 non-null   object 
 3   age_group                4030 non-null   object 
 4   race                     4030 non-null   object 
 5   gender                   4030 non-null   object 
 6   disability_count         4014 non-null   float64
 7   percent_of_total         4014 non-null   float64
 8   percent_of_disabilities  3401 non-null   float64
dtypes: float64(3), int64(1), object(5)
memory usage: 283.5+ KB


In [28]:
#Check for mixed-type data
for col in prev.columns.tolist():
    weird = (prev[[col]].applymap(type) != prev[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (prev[weird]) > 0:
        print (col)

In [29]:
prev['percent_of_disabilities']=prev['percent_of_disabilities'].astype('str')

In [30]:
#Check for mixed-type data
for col in prev.columns.tolist():
    weird = (prev[[col]].applymap(type) != prev[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (prev[weird]) > 0:
        print (col)

In [31]:
#Check Data Types in employment & poverty dataset
employ.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 714 entries, 0 to 713
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   714 non-null    object 
 1   total_pop               714 non-null    int64  
 2   disability_category     714 non-null    object 
 3   race                    714 non-null    object 
 4   gender                  714 non-null    object 
 5   employed_count          710 non-null    float64
 6   disability_employed     709 non-null    float64
 7   no_disability_employed  408 non-null    float64
 8   employment_gap          403 non-null    float64
 9   poverty_count           710 non-null    float64
 10  disability_poverty      710 non-null    float64
 11  no_disability_poverty   408 non-null    float64
 12  poverty_gap             404 non-null    float64
dtypes: float64(8), int64(1), object(4)
memory usage: 72.6+ KB


In [32]:
#Check Data Types in earnings dataset
earn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   51 non-null     object 
 1   total_earnings          51 non-null     float64
 2   no_disability_earnings  51 non-null     float64
 3   disability_earnings     51 non-null     float64
 4   earnings_gap            51 non-null     float64
 5   SSDI_allowance_rate     51 non-null     float64
 6   min_wage_rate           51 non-null     float64
dtypes: float64(6), object(1)
memory usage: 2.9+ KB


## 4. Duplicates

In [33]:
#Checking for duplicates in disability prevalence data set
prev_dups = prev[prev.duplicated()]
prev_dups

Empty DataFrame
Columns: [state, total_pop, disability_category, age_group, race, gender, disability_count, percent_of_total, percent_of_disabilities]
Index: []

In [34]:
#Checking for duplicates in employment & poverty data set
employ_dups = employ[employ.duplicated()]
employ_dups

Empty DataFrame
Columns: [state, total_pop, disability_category, race, gender, employed_count, disability_employed, no_disability_employed, employment_gap, poverty_count, disability_poverty, no_disability_poverty, poverty_gap]
Index: []

In [35]:
#Checking for duplicates in earnings data set
earn_dups = earn[earn.duplicated()]
earn_dups

Empty DataFrame
Columns: [state, total_earnings, no_disability_earnings, disability_earnings, earnings_gap, SSDI_allowance_rate, min_wage_rate]
Index: []

There are no duplicates in any of the data sets.

In [38]:
earn.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'earn2.csv'))